core/wasm/src/runtime.rs

```diff
@@ -73,6 +79,27 @@ impl<'a> Runtime<'a> {
         }
     }
 
+    fn charge_mana(&mut self, mana: u32) -> bool {
+        let prev = self.mana_counter;
+        match prev.checked_add(mana) {
+            // mana charge overflow protection
+            None => false,
```

Comment:
The first two cases can be combined by `None | Some(val) if val > self.mana_limit`

Reply:
This way is easier to read